frontend/src/metabase-lib/lib/queries/utils/structured-query-table.ts

```diff
@@ -0,0 +1,76 @@
+import type Field from "metabase-lib/lib/metadata/Field";
+import type Table from "metabase-lib/lib/metadata/Table";
+import type { Field as FieldRef } from "metabase-types/types/Query";
+
+import { isVirtualCardId } from "metabase/lib/saved-questions";
+
+import type StructuredQuery from "../StructuredQuery";
+import { createVirtualTable, createVirtualField } from "./virtual-table";
+import { getDatasetTable, getNestedCardTable } from "./nested-card-query-table";
+
+export function getStructuredQueryTable(query: StructuredQuery): Table | null {
+  const sourceQuery = query.sourceQuery();
+  // 1. Query has a source query. Use the source query as a table.
+  if (sourceQuery) {
+    return getSourceQueryTable(query);
+  }
+
+  // 2. Query has a source table that is a nested card.
+  const sourceTableId = query.sourceTableId();
+  if (isVirtualCardId(sourceTableId)) {
+    return getNestedCardTable(query);
+  }
+
+  // 3. The query's question is a dataset.
+  const question = query.question();
+  const isDataset = question?.isDataset() ?? false;
+  if (isDataset) {
+    return getDatasetTable(query);
+  }
+
+  // 4. The query's table is a concrete table, assuming one exists in `metadata`.
+  // Faiure to find a table at this point indicates that there is a bug.
+  return query.metadata().table(sourceTableId);
```

Comment:
Should we try/catch and throw if no table is found?

Reply:
For now, I'd rather not be super strict with regards to this sort of thing in terms of throwing errors, etc. Code's too complicated and there might be parts of the app the operate even though `query.table()` is `null`. If they don't work, they will surely error anyways.